text/0813-template-layering.md

```diff
@@ -0,0 +1,984 @@
+edge cases do not create a lot of additional implementation complexity.
+
+## Alternatives
+
+### Class Decorators
+
+We could use class decorators to associate templates to classes instead of the
+static initializer block approach. The biggest drawback is that we would not be
+able to access private fields with this design. In addition, in practice it
+feels pretty awkward to use due to the length of the source text and the amount
+of arguments needed, even for trivial examples. It also doesn't help this case
+that stage 3 decorators require that class decorators come after the `export`
+and `default` keywords:
+
+```js
+export @template(`<Hello />`, () => ({ Hello }) class Bar {
+}
+```
+
+### Reserving Space For "Attributes"
+
+It seems likely that the `<template>` tag will eventually gain some kind of
+"attributes" syntax for configuring things like whitespace handling (see the
+next section).
+
+One option is to "reserve space" for that now, but it can also be done later by
+optionally allowing the current "scope closure" argument to be a options object
+where the scope closure would be passed as `{ scope: () => ({ ... }) }`.
+
+## Unresolved questions
+
+### Whitespace Handling
+
+One thing that [RFC #779][rfc-779] did _not_ specify is that how whitespace
+should be handled. The default conclusion is probably that the whitespaces
+should be preserved exactly the way they appear in the `.gjs` source text.
+While this technically works, it is a conclusion that pleases and benefits no
+one, and will likely be revised/refined during the "spec work" process before
+the feature is shipped.
+
+On the one hand, there is the argument that whitespaces are typically not
+significant in almost all HTML contexts templates. Shipping all these useless
+whitespace just to have the browser fold them away when rendered into the DOM
+is a not insignificant amount of wasted space in app bundles today. It also
+just creates unnecessary work for the browser and the rendering layer.
+
+On the other hand, there is the argument that whitespaces _can_ be significant.
+Sometimes, this can be observed from uses of the `<pre>` element, but really,
+the whitespace handling for any element (including `<pre>`) can be changed
+using the CSS `white-space` property, so it is not necessarily safe to assume
+that we can fold/collapse whitespaces in templates just because that's the
+default behavior of the browser.
+
+However, even with that in mind, the default conclusion for the whitespace
+handling in `<template>` tags is _still_ not useful, _especially_ when the
+author is trying to preserve significant whitespace for something like code
+samples. This is because `<template>` tags often appears in JavaScript
+contexts that already has leading indentation (such as inside a class body), so
+in practice, if we preserve the whitespace exactly as it appears inside the
+`<template>` tag, it still would not do what the author was trying to
+accomplish.
```

Comment:
Strongly agree, and in line with the discussions we've had offline, I think we should probably apply the `HEREDOC`/etc. rules: *leading* indentation up to the indentation of the line on which the opening `<template>` appears should be stripped.